modules/objdetect/src/aruco/aruco_board.cpp

```diff
@@ -368,25 +367,29 @@ void CharucoBoardImpl::calcNearestMarkerCorners() {
     }
 }
 
-void CharucoBoardImpl::matchImagePoints(InputArrayOfArrays detectedCorners, InputArray detectedIds,
-                                   OutputArray _objPoints, OutputArray imgPoints) const {
-    if (detectedCorners.kind() == _InputArray::STD_VECTOR_VECTOR ||
-        detectedCorners.isMatVector() || detectedCorners.isUMatVector())
-        Board::Impl::matchImagePoints(detectedCorners, detectedIds, _objPoints, imgPoints);
-    else {
-        CV_Assert(detectedCorners.isMat() || detectedCorners.isVector());
-        size_t nDetected = detectedCorners.total();
-        vector<Point3f> objPnts(nDetected);
-        vector<Point2f> imgPnts(nDetected);
-        for(size_t i = 0ull; i < nDetected; i++) {
-            int pointId = detectedIds.getMat().at<int>((int)i);
-            CV_Assert(pointId >= 0 && pointId < (int)chessboardCorners.size());
-            objPnts[i] = chessboardCorners[pointId];
-            imgPnts[i] = detectedCorners.getMat().at<Point2f>((int)i);
+void CharucoBoardImpl::matchImagePoints(InputArrayOfArrays detectedCharuco, InputArray detectedIds,
+                                        OutputArray outObjPoints, OutputArray outImgPoints) const {
+    CV_Assert(detectedIds.total() == detectedCharuco.total());
+    CV_Assert(detectedIds.total() > 0ull);
+    // detectedCharuco includes charuco corners as vector<Point2f> or Mat.
+    // Python bindings could add extra dimension to detectedCharuco and therefore vector<Mat> case is additionally rocessed.
+    CV_Assert(detectedCharuco.isMat() || detectedCharuco.isVector() || detectedCharuco.isMatVector());
+    size_t nDetected = detectedCharuco.total();
+    vector<Point3f> objPnts(nDetected);
+    vector<Point2f> imgPnts(nDetected);
+    for(size_t i = 0ull; i < nDetected; i++) {
+        int pointId = detectedIds.getMat().at<int>((int)i);
+        CV_Assert(pointId >= 0 && pointId < (int)chessboardCorners.size());
+        objPnts[i] = chessboardCorners[pointId];
+        if (detectedCharuco.isMatVector()) { // python binding case
+            imgPnts[i].x = detectedCharuco.getMat((int)i).at<float>(0);
+            imgPnts[i].y = detectedCharuco.getMat((int)i).at<float>(1);
         }
-        Mat(objPnts).copyTo(_objPoints);
-        Mat(imgPnts).copyTo(imgPoints);
+        else
+            imgPnts[i] = detectedCharuco.getMat().at<Point2f>((int)i);
```

Comment:
> `.at<Point2f>`

Missing type checks.

Reply:
add Assert: CV_Assert(... && detectedCharuco.depth() == CV_32F)
add Assert: CV_Assert((int)detectedCharucoMat.channels() == 2) for C++ case
add Assert: CV_Assert((int)detectedCharucoVecMat.front().total() * detectedCharucoVecMat.front().channels() == 2) for python binding case